redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,639 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+
+use crate::aio::ConnectionLike;
+use crate::cluster_async::{ClusterConnInner, Connect, InnerCore};
+use crate::cluster_routing::SlotAddr;
+use crate::{cmd, from_redis_value, Cmd, ErrorKind, RedisError, RedisResult, Value};
+
+const BITS_PER_U64: usize = 64;
+const NUM_OF_SLOTS: usize = 16384;
+const BITS_ARRAY_SIZE: usize = NUM_OF_SLOTS / BITS_PER_U64;
+const END_OF_SCAN: u16 = NUM_OF_SLOTS as u16 + 1;
+type SlotsBitsArray = [u64; BITS_ARRAY_SIZE];
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct ScanState {
+    pub cursor: u64,
+    pub scanned_slots_map: SlotsBitsArray,
+    pub slots_in_scan_map: SlotsBitsArray,
+    pub node_in_scan: String,
+    pub node_epoch: u64,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ClusterScanArgs {
+    scan_state_ref: ScanStateWrapper,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<String>,
+}
+
+impl ClusterScanArgs {
+    pub(crate) fn new(
+        scan_state_ref: ScanStateWrapper,
+        match_pattern: Option<String>,
+        count: Option<usize>,
+        object_type: Option<String>,
+    ) -> Self {
+        Self {
+            scan_state_ref,
+            match_pattern,
+            count,
+            object_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// A wrapper struct for managing the state of a scan operation in a cluster.
+pub struct ScanStateWrapper {
+    scan_atate_ref: Arc<Option<ScanState>>,
+}
+
+impl PartialEq for ScanStateWrapper {
+    fn eq(&self, other: &Self) -> bool {
+        self.is_none() == other.is_none()
+    }
+}
+
+impl Default for ScanStateWrapper {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ScanStateWrapper {
+    fn from(scan_state: &ScanState) -> Self {
+        Self {
+            scan_atate_ref: Arc::new(Some(scan_state.clone())),
+        }
+    }
+
+    /// Creates a new instance of `ScanStateWrapper`.
+    ///
+    /// This method initializes the `ScanStateWrapper` with a read-write lock
+    /// that is initially set to `None`.
+    pub fn new() -> Self {
+        Self {
+            scan_atate_ref: Arc::new(None),
+        }
+    }
+
+    /// Returns `true` if the scan state is empty, indicating that the scan operation has finished or has not started.
+    pub fn is_none(&self) -> bool {
+        self.scan_atate_ref.is_none()
+    }
+
+    /// Returns a clone of the scan state, if it exists.
+    pub fn get_state_from_wrrraper(&self) -> ScanState {
+        self.scan_atate_ref.as_ref().clone().unwrap()
+    }
+}
+
+/// This trait defines the methods for interacting with a Redis cluster during scanning.
+#[async_trait]
+pub(crate) trait ClusterInScan {
+    /// Retrieves the node address associated with a given slot in the cluster.
+    async fn get_node_by_slot(&self, slot: u16) -> String;
+
+    /// Retrieves the epoch of a given node in the cluster.
+    async fn get_node_epoch(&self, node: &str) -> Result<u64, RedisError>;
```

Comment:
the syntax of `node` in the connections container / async cluster is that node is the struct that contains the connections, and what you're searching for is the `address`